src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/RazorLSPMappingService.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Protocol.DocumentMapping;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Razor.LanguageClient.DocumentMapping;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient;
+
+internal sealed class RazorLSPMappingService(
+    LSPDocumentMappingProvider lspDocumentMappingProvider,
+    LSPDocumentSnapshot documentSnapshot,
+    ITextSnapshot textSnapshot) : IRazorMappingService
+{
+    private readonly LSPDocumentMappingProvider _lspDocumentMappingProvider = lspDocumentMappingProvider;
+    private readonly LSPDocumentSnapshot _documentSnapshot = documentSnapshot;
+    private readonly ITextSnapshot _textSnapshot = textSnapshot;
+
+    public async Task<ImmutableArray<RazorMappedSpanResult>> MapSpansAsync(
+        Document document,
+        IEnumerable<TextSpan> spans,
+        CancellationToken cancellationToken)
+    {
+        return await MapSpansAsync(
+            spans,
+            _textSnapshot.AsText(),
+            _documentSnapshot.Snapshot.AsText(),
+            cancellationToken).ConfigureAwait(false);
+    }
+
+    public async Task<ImmutableArray<RazorMappedEditoResult>> MapTextChangesAsync(
+        Document oldDocument,
+        Document newDocument,
+        CancellationToken cancellationToken)
+    {
+        var changes = await newDocument.GetTextChangesAsync(oldDocument, cancellationToken).ConfigureAwait(false);
+
+        var mappedEdits = await _lspDocumentMappingProvider.MapToDocumentEditssAsync(
+            RazorLanguageKind.CSharp,
+            _documentSnapshot.Uri,
+            changes.ToArray(),
+            cancellationToken);
+
+        if (mappedEdits is null)
+        {
+            return [];
+        }
+
+        return [new RazorMappedEditoResult(_documentSnapshot.Uri.AbsolutePath, mappedEdits.Edits)];
+    }
+
+    private async Task<ImmutableArray<RazorMappedSpanResult>> MapSpansAsync(
+        IEnumerable<TextSpan> spans,
+        SourceText sourceTextGenerated,
+        SourceText sourceTextRazor,
+        CancellationToken cancellationToken)
+    {
+        if (spans is null)
+        {
+            throw new ArgumentNullException(nameof(spans));
+        }
```

Comment:
This one is definitely a "Draft, look back over later" state because all of this was directly copied from the existing implementation :) I have a note to go back and do cleanup

Reply:
Makes sense.